ethereum/dbstack/db_stack.go

```diff
@@ -1,91 +1,168 @@
+		meshDB:      meshDB,
+		limit:       retentionLimit,
+		miniHeaders: miniHeaders,
+		updates:     []*update{},
+	}
+	return d, nil
+}
+
+// Peek returns the top of the stack
+func (d *DBStack) Peek() (*miniheader.MiniHeader, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if len(d.miniHeaders) == 0 {
 		return nil, nil
 	}
-	if err := b.meshDB.MiniHeaders.Delete(latestMiniHeader.ID()); err != nil {
-		return nil, err
+	return d.miniHeaders[len(d.miniHeaders)-1], nil
+}
+
+// Pop returns the top of the stack and removes it from the stack
+func (d *DBStack) Pop() (*miniheader.MiniHeader, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if len(d.miniHeaders) == 0 {
+		return nil, nil
 	}
-	return latestMiniHeader, nil
+	top := d.miniHeaders[len(d.miniHeaders)-1]
+	d.miniHeaders = d.miniHeaders[:len(d.miniHeaders)-1]
+	d.updates = append(d.updates, &update{
+		Type: pop,
+		// Optimization: We don't need to store the MiniHeader for
+		// pops since checkpoints a pop involves removing the latest
+		// block header from the DB which doesn't require the explicit
+		// header value
+		MiniHeader: nil,
+	})
+	return top, nil
 }
 
-// Push pushes a block header onto the block stack. If the stack limit is
-// reached, it will remove the oldest block header.
-func (b *DBStack) Push(block *miniheader.MiniHeader) error {
-	b.mut.Lock()
-	defer b.mut.Unlock()
-	miniHeaders, err := b.meshDB.FindAllMiniHeadersSortedByNumber()
-	if err != nil {
-		return err
+// Push adds a miniheader.MiniHeader to the stack
+func (d *DBStack) Push(miniHeader *miniheader.MiniHeader) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if len(d.miniHeaders) == d.limit {
+		d.miniHeaders = d.miniHeaders[1:]
 	}
-	if len(miniHeaders) == b.limit {
-		oldestMiniHeader := miniHeaders[0]
-		if err := b.meshDB.MiniHeaders.Delete(oldestMiniHeader.ID()); err != nil {
-			return err
+	d.miniHeaders = append(d.miniHeaders, miniHeader)
+	d.updates = append(d.updates, &update{
+		Type:       push,
+		MiniHeader: miniHeader,
+	})
+	return nil
+}
+
+// PeekAll returns all the miniHeaders currently in the stack
+func (d *DBStack) PeekAll() ([]*miniheader.MiniHeader, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	return d.miniHeaders, nil
+}
+
+// Clear removes all items from the stack
+func (d *DBStack) Clear() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.miniHeaders = []*miniheader.MiniHeader{}
+	return d.meshDB.ClearAllMiniHeaders()
+}
+
+// Checkpoint checkpoints the changes to the stack such that a subsequent
+// call to `Reset()` will reset any subsequent changes back to the state
+// of the stack at the time of the latest checkpoint. The checkpointed state
+// is also persisted to the DB.
+func (d *DBStack) Checkpoint() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	txn := d.meshDB.MiniHeaders.OpenTransaction()
+	defer func() {
+		_ = txn.Discard()
+	}()
+
+	for _, u := range d.updates {
+		switch u.Type {
+		case pop:
+			latestMiniHeader, err := d.meshDB.FindLatestMiniHeader()
```

Comment:
In the context of a `db.Transaction`, nothing is actually written to the database until `Commit` is called. That means that `FindLatestMiniHeader` always returns the same `MiniHeader`, whatever the latests is at the time the transaction was opened. I don't think this is what you intended.

If I'm right and this is a bug, we should also improve our test coverage so we can catch cases like this.

Reply:
Good catch. Fixed and added regression test.